plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostModelInterface.hs

```diff
@@ -0,0 +1,159 @@
+costs of the evaluator and a model for built-in evaluation costs.  We just
+re-use the technique mentioned above to extract parameters for the evaluator
+costs, merging these with the parameters for the builtin cost model to obtain
+parameters for the overall model.  To recover cost model components we assume
+that every field in the cost model for the evaluator begins with a prefix (eg
+"cek") which is does not occur as a prefix of any built-in function, and use
+that to split the map of parameters into two maps.
+
+-}
+
+-- | See point 3 of Note [Cost model parameters]
+-- Some of the numbers in the cost model are of the order or 1e-7 or 1e-8, and
+-- we need a large scale factor to avoid truncating a significant number of
+-- digits.
+scaleFactor :: S.Scientific
+scaleFactor = 1e20
+
+-- See Note [Cost model parameters]
+type CostModelParams = Map.Map Text.Text Integer
+
+-- See Note [Cost model parameters]
+-- | Extract the model parameters from a model.
+extractParams :: ToJSON a => a -> Maybe CostModelParams
+extractParams cm = case toJSON cm of
+    Object o ->
+        let
+            flattened = flattenObject "-" o
+            toScaledInteger :: S.Scientific -> Integer
+            toScaledInteger n = floor (n*scaleFactor)
+            scaledNumbers = HM.mapMaybe (\case { Number n -> Just $ toScaledInteger n; _ -> Nothing }) flattened
+            mapified = Map.fromList $ HM.toList scaledNumbers
+        in Just mapified
+    _ -> Nothing
+
+
+-- See Note [Cost model parameters]
+-- | Update a model by overwriting the parameters with the given ones.
+applyParams :: (FromJSON a, ToJSON a) => a -> CostModelParams -> Maybe a
+applyParams cm params = case toJSON cm of
+    Object o ->
+        let
+            hashmapified = HM.fromList $ Map.toList params
+            scaledNumbers = fmap (\n -> Number $ fromIntegral n / scaleFactor) hashmapified
+            flattened = flattenObject "-" o
+            -- this is where the overwriting happens, this is left-biased
+            merged = HM.union scaledNumbers flattened
+            unflattened = unflattenObject "-" merged
+        in case fromJSON (Object unflattened) of
+            Success a -> Just a
+            Error _   -> Nothing
+    _ -> Nothing
+
+-- | Parameters for a machine step model and a builtin evaluation model bundled together.
+data SplitCostModelParams =
+    SplitCostModelParams {
+      machineParams :: CostModelParams
+    , builtinParams :: CostModelParams
+    }
+
+-- | Split a CostModelParams object into two subobjects according to some prefix:
+-- see item 5 of Note [Cost model parameters].
+splitParams :: Text.Text -> CostModelParams -> SplitCostModelParams
+splitParams prefix params =
+    let machineparams = Map.filterWithKey (\k _ ->       Text.isPrefixOf prefix k) params
+        builtinparams = Map.filterWithKey (\k _ -> not $ Text.isPrefixOf prefix k) params
+    in SplitCostModelParams machineparams builtinparams
+
```

Comment:
Perhaps we don't need to do this splitting.  Experiments suggest that for simple types at least you can decode from JSON objects that contain superfluous components: for example, if you've got `Point = Point {x::Int, y::Int}` then `fromJSON@Point` will decode any JSON object that contains keys `x` and `y`, even if it also has other keys like `z` and `t`.  If we were willing to rely on this then we could overwrite the JSON encodings  of both the machine costs and he builtin costs with the full set of params: this would introduce extra fields, but these would be discarded when we decoded back to Haskell objects. 

However I wasn't able to see this behaviour described anywhere in the Aeson documentation, so it'd probably be unwise to rely on it.  This code's already a bit of a hack, and making it depend on undocumented behaviour wouldn't make things better.